tfhe/src/integer/gpu/mod.rs

```diff
@@ -273,6 +276,160 @@ pub unsafe fn unchecked_scalar_mul_integer_radix_kb_async<T: UnsignedInteger, B:
+    carry_modulus: CarryModulus,
+    compression_glwe_dimension: GlweDimension,
+    compression_polynomial_size: PolynomialSize,
+    lwe_dimension: LweDimension,
+    ks_base_log: DecompositionBaseLog,
+    ks_level: DecompositionLevelCount,
+    lwe_per_glwe: u32,
+    storage_log_modulus: u32,
+    num_blocks: u32,
+) {
+    assert_eq!(
+        streams.gpu_indexes[0],
+        glwe_array_out.gpu_index(0),
+        "GPU error: all data should reside on the same GPU."
+    );
+    assert_eq!(
+        streams.gpu_indexes[0],
+        lwe_array_in.gpu_index(0),
+        "GPU error: all data should reside on the same GPU."
+    );
+    assert_eq!(
+        streams.gpu_indexes[0],
+        fp_keyswitch_key.gpu_index(0),
+        "GPU error: all data should reside on the same GPU."
+    );
+    let mut mem_ptr: *mut i8 = std::ptr::null_mut();
+    scratch_cuda_integer_compress_radix_ciphertext_64(
+        streams.ptr.as_ptr(),
+        streams.gpu_indexes.as_ptr(),
+        streams.len() as u32,
+        std::ptr::addr_of_mut!(mem_ptr),
+        compression_glwe_dimension.0 as u32,
+        compression_polynomial_size.0 as u32,
+        lwe_dimension.0 as u32,
+        ks_level.0 as u32,
+        ks_base_log.0 as u32,
+        num_blocks,
+        message_modulus.0 as u32,
+        carry_modulus.0 as u32,
+        PBSType::Classical as u32,
+        lwe_per_glwe,
+        storage_log_modulus,
+        true,
+    );
+
+    cuda_integer_compress_radix_ciphertext_64(
+        streams.ptr.as_ptr(),
+        streams.gpu_indexes.as_ptr(),
+        streams.len() as u32,
+        glwe_array_out.as_mut_c_ptr(0),
+        lwe_array_in.as_c_ptr(0),
+        fp_keyswitch_key.ptr.as_ptr(),
+        num_blocks,
+        mem_ptr,
+    );
+
+    cleanup_cuda_integer_compress_radix_ciphertext_64(
+        streams.ptr.as_ptr(),
+        streams.gpu_indexes.as_ptr(),
+        streams.len() as u32,
+        std::ptr::addr_of_mut!(mem_ptr),
+    );
+}
+
+#[allow(clippy::too_many_arguments)]
+/// # Safety
+///
+/// - [CudaStreams::synchronize] __must__ be called after this function as soon as synchronization
+///   is required
+pub unsafe fn decompress_integer_radix_async<T: UnsignedInteger, B: Numeric>(
+    streams: &CudaStreams,
+    lwe_array_out: &mut CudaVec<T>,
+    glwe_in: &CudaVec<T>,
+    bootstrapping_key: &CudaVec<B>,
+    message_modulus: MessageModulus,
+    carry_modulus: CarryModulus,
+    encryption_glwe_dimension: GlweDimension,
+    encryption_polynomial_size: PolynomialSize,
+    compression_glwe_dimension: GlweDimension,
+    compression_polynomial_size: PolynomialSize,
+    lwe_dimension: LweDimension,
+    pbs_base_log: DecompositionBaseLog,
+    pbs_level: DecompositionLevelCount,
+    storage_log_modulus: u32,
+    vec_indexes: &CudaVec<u32>,
+    num_blocks: u32,
+) {
+    assert_eq!(
+        streams.gpu_indexes[0],
+        lwe_array_out.gpu_index(0),
+        "GPU error: all data should reside on the same GPU."
+    );
```

Comment:
same questions about the checks on the first element of the array only, I'm not sure I understand why

Reply:
`streams` stores a list of `cudaStream_t`, each one related to a different device (if multi-gpu is available, of course). This list does not need to contain all N available devices, does not need to start at 0, and does not need to be sequential. So, `1,4,5` is a possibility on a machine with 8 devices.

`CudaVec` stores a list of pointers, each one related to a different device. In the future we may use that in different ways, but today we only store data in `gpu_index[0]`.  

If user is executing their code on a single-GPU environment, there is only one GPU possibility and this check is useless. 

If user is executing their code on a multi-GPU environment, stream in `stream.gpu_indexes[0]` will be the "main" stream, and all kernels that do not support multi-gpu computation will be launched on it. We assume all inputs and outputs are read or written to the GPU memory that is related to that stream. Thus, in this code snippet, we need `streams.gpu_indexes[0] == lwe_array_out.gpu_index(0)`. Internally, we have temporary arrays in different devices that will be used in the multi-GPU logic, but everything starts and ends in `gpu_index[0]`, whatever it is.

Reply:
thanks for the explanation, as a side note not sure CudaVec is the right name for the primitive anymore, as a vec is generally a contiguous growable array of data in memory, feels like here it's likely not contiguous in a multi GPU setting